File: Daily/645. Set Mismatch/Solution1.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;
class Solution {
    public:
        vector<int> findErrorNums(vector<int>& nums) {
            /*
                Optimized Solution using Sorting
                Time Complexity = O(nlogn)
                Space Complexity = O(logn)
            */
            sort(nums.begin(), nums.end());
            int duplicate = -1, missing = 1; const int size = nums.size();
            vector<int> error(2);
            for(int index = 1; index < size; index++) {
                if(nums[index - 1] == nums[index]) duplicate = nums[index];
                else if(nums[index] > nums[index - 1] + 1) missing = nums[index - 1] + 1;
            }
            missing = (nums[size - 1] != size) ? size : missing;
            error[0] = duplicate; error[1] = missing;
            return error;
        }
};
int main() {
    int n; cin >> n; vector<int> nums(n); Solution sol;
    for(int index = 0; index < n; ++index) cin >> nums[index]; 
    vector<int> ans = sol.findErrorNums(nums);
    cout << "Duplicate Element = " << ans[0] 
        << " and Missing Element = " << ans[1] << endl;
}

File: Daily/645. Set Mismatch/Solution1.py
from typing import List
class Solution:
    def findErrorNums(self, nums: List[int]) -> List[int]:
        '''
            Optimized Solution using Sorting
            Time Complexity = O(nlogn)
            Space Complexity = O(logn)
        '''
        duplicate, missing, size = -1, 1, len(nums)
        nums.sort()
        for index in range(1, size):
            if nums[index] == nums[index - 1]: duplicate = nums[index]
            if nums[index] != nums[index - 1] + 1: missing = nums[index - 1] + 1
        missing = size if nums[size - 1] != size else missing
        return [duplicate, missing]

if __name__ == '__main__':
    nums = [int(x) for x in input().split()]; sol = Solution()
    ans = sol.findErrorNums(nums)
    print('Duplicate Number =', ans[0], 'Missing Number =', ans[1])